prow/plugins/approve/approve.go

```diff
@@ -0,0 +1,508 @@
+	commandRegex         = regexp.MustCompile(`(?m)^/([^\s]+)[\t ]*([^\n\r]*)`)
+	notificationRegex    = regexp.MustCompile(`(?is)^\[` + approvers.ApprovalNotificationName + `\] *?([^\n]*)(?:\n\n(.*))?`)
+)
+
+type githubClient interface {
+	GetPullRequestChanges(org, repo string, number int) ([]github.PullRequestChange, error)
+	GetIssueLabels(org, repo string, number int) ([]github.Label, error)
+	ListIssueComments(org, repo string, number int) ([]github.IssueComment, error)
+	ListReviews(org, repo string, number int) ([]github.Review, error)
+	ListPullRequestComments(org, repo string, number int) ([]github.ReviewComment, error)
+	DeleteComment(org, repo string, ID int) error
+	CreateComment(org, repo string, number int, comment string) error
+	BotName() (string, error)
+	AddLabel(org, repo string, number int, label string) error
+	RemoveLabel(org, repo string, number int, label string) error
+	ListIssueEvents(org, repo string, num int) ([]github.ListedIssueEvent, error)
+}
+
+type state struct {
+	org    string
+	repo   string
+	number int
+
+	body      string
+	author    string
+	assignees []github.User
+	htmlURL   string
+
+	repoOptions *plugins.Approve
+}
+
+func handleGenericCommentEvent(pc plugins.PluginClient, ce github.GenericCommentEvent) error {
+	ro, err := pc.OwnersClient.LoadRepoOwners(ce.Repo.Owner.Login, ce.Repo.Name)
+	if err != nil {
+		return err
+	}
+	return handleGenericComment(pc.Logger, pc.GitHubClient, ro, pc.PluginConfig, &ce)
+}
+
+func handleGenericComment(log *logrus.Entry, ghc githubClient, repo approvers.RepoInterface, config *plugins.Configuration, ce *github.GenericCommentEvent) error {
+	if ce.Action != github.GenericCommentActionCreated || !ce.IsPR {
+		return nil
+	}
+
+	botName, err := ghc.BotName()
+	if err != nil {
+		return err
+	}
+
+	if !approvalCommandMatcher(botName)(&comment{Body: ce.Body, Author: ce.User.Login}) {
+		return nil
+	}
+
+	return handle(
+		log,
+		ghc,
+		repo,
+		optionsForRepo(config, ce.Repo.Owner.Login, ce.Repo.Name),
+		&state{
+			org:       ce.Repo.Owner.Login,
+			repo:      ce.Repo.Name,
+			number:    ce.Number,
+			body:      ce.IssueBody,
+			author:    ce.IssueAuthor.Login,
+			assignees: ce.Assignees,
+			htmlURL:   ce.IssueHTMLURL,
+		},
+	)
+}
+
+func handlePullRequestEvent(pc plugins.PluginClient, pre github.PullRequestEvent) error {
+	ro, err := pc.OwnersClient.LoadRepoOwners(pre.Repo.Owner.Login, pre.Repo.Name)
+	if err != nil {
+		return err
+	}
+	return handlePullRequest(pc.Logger, pc.GitHubClient, ro, pc.PluginConfig, &pre)
+}
+
+func handlePullRequest(log *logrus.Entry, ghc githubClient, repo approvers.RepoInterface, config *plugins.Configuration, pre *github.PullRequestEvent) error {
+	if pre.Action != github.PullRequestActionOpened &&
+		pre.Action != github.PullRequestActionReopened &&
+		pre.Action != github.PullRequestActionSynchronize &&
+		pre.Action != github.PullRequestActionLabeled {
+		return nil
+	}
+
+	botName, err := ghc.BotName()
```

Comment:
There are a bunch of plugins that need this so we could potentially figure this out on hook startup once, cache, and then reuse it. Follow-up issue to discuss is better than fixing it here.

Reply:
The github client already does that. :smile:

Reply:
Yeah, figured that out in https://github.com/kubernetes/test-infra/pull/5249 - nevermind